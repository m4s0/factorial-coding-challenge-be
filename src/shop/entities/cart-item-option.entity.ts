import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { CartItem } from './cart-item.entity';
import { ProductOption } from './product-option.entity';

@Entity()
export class CartItemOption {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @ManyToOne(() => CartItem, (cartItem) => cartItem.itemOptions, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare cartItem: CartItem;

  @Column()
  declare cartItemId: string;

  @ManyToOne(() => ProductOption, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare option: ProductOption;

  @Column()
  declare optionId: string;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
