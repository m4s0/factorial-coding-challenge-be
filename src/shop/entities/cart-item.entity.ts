import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { CartItemOption } from '@Shop/entities/cart-item-option.entity';
import { Cart } from './cart.entity';
import { Product } from './product.entity';

@Entity()
export class CartItem {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @ManyToOne(() => Cart, (cart) => cart.items, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare cart: Cart;

  @Column()
  declare cartId: string;

  @ManyToOne(() => Product, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare product: Product;

  @Column()
  declare productId: string;

  @Column({ type: 'int', default: 1 })
  declare quantity: number;

  @OneToMany(
    () => CartItemOption,
    (cartItemOption) => cartItemOption.cartItem,
    {
      cascade: true,
    },
  )
  declare itemOptions: CartItemOption[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare price: number;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare totalPrice: number;
}
