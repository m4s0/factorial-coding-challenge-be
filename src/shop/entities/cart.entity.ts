import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { User } from '@User/entities/user.entity';
import { CartItem } from './cart-item.entity';

@Entity()
export class Cart {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @OneToOne(() => User, (user) => user.cart, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare user: User;

  @Column()
  declare userId: string;

  @OneToMany(() => CartItem, (item) => item.cart, {
    cascade: true,
  })
  declare items: CartItem[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;

  declare totalPrice?: number;
}
