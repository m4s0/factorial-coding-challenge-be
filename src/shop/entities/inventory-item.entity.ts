import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { ProductOption } from './product-option.entity';

@Entity()
export class InventoryItem {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({ type: 'int' })
  declare quantity: number;

  @Column({ default: false })
  declare outOfStock: boolean;

  @OneToOne(
    () => ProductOption,
    (productOption) => productOption.inventoryItem,
    {
      onDelete: 'CASCADE',
    },
  )
  @JoinColumn()
  declare productOption: ProductOption;

  @Column()
  declare productOptionId: string;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
