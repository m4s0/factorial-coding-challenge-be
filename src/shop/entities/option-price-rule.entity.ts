import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { ProductOption } from './product-option.entity';

@Entity()
export class OptionPriceRule {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare price: number;

  @ManyToOne(() => ProductOption, (option) => option.priceRules, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare targetOption: ProductOption;

  @Column()
  declare targetOptionId: string;

  @ManyToOne(() => ProductOption, (option) => option.priceRules, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare dependentOption: ProductOption;

  @Column()
  declare dependentOptionId: string;

  @Column({ default: true })
  declare isActive: boolean;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
