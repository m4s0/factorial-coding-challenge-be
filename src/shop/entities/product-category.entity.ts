import {
  Column,
  CreateDateColumn,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { Product } from './product.entity';

@Entity()
export class ProductCategory {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column()
  declare name: string;

  @Column()
  declare description: string;

  @Column({ default: true })
  declare isActive: boolean;

  @OneToMany(() => Product, (product) => product.category)
  declare products: Product[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
