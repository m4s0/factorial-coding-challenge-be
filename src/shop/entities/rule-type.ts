export enum RuleType {
  REQUIRES = 'requires',
  EXCLUDES = 'excludes',
  ONLY_ALLOWS = 'only_allows',
}
