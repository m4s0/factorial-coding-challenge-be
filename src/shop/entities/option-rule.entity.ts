import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { RuleType } from './rule-type';
import { ProductOption } from './product-option.entity';

@Entity()
export class OptionRule {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({
    type: 'enum',
    enum: RuleType,
  })
  declare ruleType: RuleType;

  @ManyToOne(() => ProductOption, (option) => option.rulesAsCondition, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare ifOption: ProductOption;

  @Column()
  declare ifOptionId: string;

  @ManyToOne(() => ProductOption, (option) => option.rulesAsResult, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare thenOption: ProductOption;

  @Column()
  declare thenOptionId: string;

  @Column({ default: true })
  declare isActive: boolean;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
