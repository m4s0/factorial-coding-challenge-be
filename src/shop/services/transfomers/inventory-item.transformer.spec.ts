import { InventoryItem } from '@Shop/entities/inventory-item.entity';
import { ProductOption } from '@Shop/entities/product-option.entity';
import { RuleType } from '@Shop/entities/rule-type';
import { OptionRule } from '@Shop/entities/option-rule.entity';
import { OptionPriceRule } from '@Shop/entities/option-price-rule.entity';
import { transformInventoryItem } from './inventory-item.transformer';

describe('transformInventoryItem', () => {
  it('should transform inventory item with all fields correctly', () => {
    const inventoryItem = {
      id: '755cf8ae-950c-4be2-9a20-1fb6576b745e',
      quantity: 100,
      outOfStock: false,
      productOptionId: '13585532-a295-4a82-b7bf-0abc07064ee9',
      productOption: {
        id: 'fb5d0e88-8014-4883-82d5-94d0af8e723f',
        name: 'Size',
        displayName: 'Product Size',
        basePrice: 10.99,
        isActive: true,
        optionGroupId: '31b7d272-3664-4a7f-87f7-09d95c60dfac',
        rulesAsCondition: [
          {
            id: '13585532-a295-4a82-b7bf-0abc07064ee9',
            ruleType: RuleType.REQUIRES,
            isActive: true,
            ifOptionId: 'a8f340f5-f724-43f8-b0c8-c43c29220923',
            thenOptionId: '31b7d272-3664-4a7f-87f7-09d95c60dfac',
            createdAt: new Date('2024-01-01'),
            updatedAt: new Date('2024-01-02'),
          } as OptionRule,
        ],
        rulesAsResult: [
          {
            id: 'd8cb0d41-f2bf-40ec-8f0d-24316c9eb94b',
            ruleType: RuleType.REQUIRES,
            isActive: true,
            ifOptionId: 'ec428b31-7b25-472f-813f-ff05022f580e',
            thenOptionId: '4f0bbaec-fe89-43d6-ac39-9fa215bd3bbd',
            createdAt: new Date('2024-01-01'),
            updatedAt: new Date('2024-01-02'),
          } as OptionRule,
        ],
        priceRules: [
          {
            id: '2b6c2ef0-776e-4e55-b3ad-7059539150ce',
            price: 5.0,
            targetOptionId: 'fb5d0e88-8014-4883-82d5-94d0af8e723f',
            dependentOptionId: 'a2444f70-749c-41e8-bdc3-2de2348afeb0',
            isActive: true,
            createdAt: new Date('2024-01-01'),
            updatedAt: new Date('2024-01-02'),
          } as OptionPriceRule,
        ],
        createdAt: new Date('2024-01-01'),
        updatedAt: new Date('2024-01-02'),
      } as ProductOption,
      createdAt: new Date('2024-01-01'),
      updatedAt: new Date('2024-01-02'),
    } as InventoryItem;

    const result = transformInventoryItem(inventoryItem);

    expect(result).toMatchSnapshot();
  });

  it('should transform inventory item with zero quantity and out of stock', () => {
    const inventoryItem = {
      id: 'a8f340f5-f724-43f8-b0c8-c43c29220923',
      quantity: 0,
      outOfStock: true,
      productOptionId: '4636127e-769e-4136-9a6a-8fddb7ca9692',
      createdAt: new Date('2024-01-01'),
      updatedAt: new Date('2024-01-02'),
    } as InventoryItem;

    const result = transformInventoryItem(inventoryItem);

    expect(result).toMatchSnapshot();
  });
});
