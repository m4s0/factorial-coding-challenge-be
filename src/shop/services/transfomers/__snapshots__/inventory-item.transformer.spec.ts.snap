// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformInventoryItem should transform inventory item with all fields correctly 1`] = `
{
  "createdAt": "2024-01-01T00:00:00.000Z",
  "id": "755cf8ae-950c-4be2-9a20-1fb6576b745e",
  "outOfStock": false,
  "productOption": {
    "basePrice": 10.99,
    "createdAt": "2024-01-01T00:00:00.000Z",
    "displayName": "Product Size",
    "id": "fb5d0e88-8014-4883-82d5-94d0af8e723f",
    "isActive": true,
    "name": "Size",
    "priceRules": [
      {
        "createdAt": "2024-01-01T00:00:00.000Z",
        "dependentOptionId": "a2444f70-749c-41e8-bdc3-2de2348afeb0",
        "id": "2b6c2ef0-776e-4e55-b3ad-7059539150ce",
        "isActive": true,
        "price": 5,
        "targetOptionId": "fb5d0e88-8014-4883-82d5-94d0af8e723f",
        "updatedAt": "2024-01-02T00:00:00.000Z",
      },
    ],
    "rulesAsCondition": [
      {
        "createdAt": "2024-01-01T00:00:00.000Z",
        "id": "13585532-a295-4a82-b7bf-0abc07064ee9",
        "ifOptionId": "a8f340f5-f724-43f8-b0c8-c43c29220923",
        "isActive": true,
        "ruleType": "requires",
        "thenOptionId": "31b7d272-3664-4a7f-87f7-09d95c60dfac",
        "updatedAt": "2024-01-02T00:00:00.000Z",
      },
    ],
    "rulesAsResult": [
      {
        "createdAt": "2024-01-01T00:00:00.000Z",
        "id": "d8cb0d41-f2bf-40ec-8f0d-24316c9eb94b",
        "ifOptionId": "ec428b31-7b25-472f-813f-ff05022f580e",
        "isActive": true,
        "ruleType": "requires",
        "thenOptionId": "4f0bbaec-fe89-43d6-ac39-9fa215bd3bbd",
        "updatedAt": "2024-01-02T00:00:00.000Z",
      },
    ],
    "updatedAt": "2024-01-02T00:00:00.000Z",
  },
  "productOptionId": "13585532-a295-4a82-b7bf-0abc07064ee9",
  "quantity": 100,
  "updatedAt": "2024-01-02T00:00:00.000Z",
}
`;

exports[`transformInventoryItem should transform inventory item with zero quantity and out of stock 1`] = `
{
  "createdAt": "2024-01-01T00:00:00.000Z",
  "id": "a8f340f5-f724-43f8-b0c8-c43c29220923",
  "outOfStock": true,
  "productOptionId": "4636127e-769e-4136-9a6a-8fddb7ca9692",
  "quantity": 0,
  "updatedAt": "2024-01-02T00:00:00.000Z",
}
`;
